### ISSUE TITLE
Provide a service for Virtual File System Registration

### ISSUE THREAD

**Issue by steelbrain** [atom/tree-view#452]
#### The Problem

Have a look at [https://atom.io/packages/remote-ftp](https://atom.io/packages/remote-ftp). There are a other packages like that that provides a similar functionality by completely forking the tree-view project and then modifying the fs functions with their own. Its not the most beautiful and elegant solution.
Other than that, there is no nicer way known to me to show remote folder and files over SFTP or FTP or network mounts in the tree-view.

#### The Solution
The best solution I can think of right now would be for tree-view to provide a service for packages to either register a `protocol` or `type`.

The FileSystem Providers would return something like this from their provide function
```js
type FileInfo = shape(Name => String, Type = enum{File, Dir}, Permissions => int, LasttModified => Date)
class TreeViewProvider{
  constructor(ConstructorParams:Object)
  directoryExists: function(Path:String):Boolean
  directoryContents: function(Path:String):Array<FileInfo>
  fileExists: function(Path:String):Boolean
  fileContents: function(Path:String):String
  rename: function(Path:String):Boolean // false on failure
  unlink: function(Path:String):Boolean // false on failure
}
```
##### Protocol
The Tree Provider will provide a `protocol:ssh` or something similar in the object it returns. Then other packages or the user or anyone who wants to open a project dir over that protocol would do something like `atom.project.open('ssh://192.168.1.10//var/www', {user: 'steel', 'password': 'mysecretstuff'})` and the extra arguments will be provided to the constructor of that provide.

##### Type
It's gonna be the same as protocol, just that it would have a field named type like `type:ssh` and the openers would do something like `atom.project.open('/var/www', {type:'ssh'}, {user: 'steel', 'password': 'mysecretstuff'})`.

__Note__: The methods mention above that `something like`. The API would be something similar to that, and not literally that. as I am kind-of sure it's already using the second and third parameters for something.

**Comment by izuzak** [comment]
Maybe related: https://github.com/atom/tree-view/issues/207 (see https://github.com/atom/tree-view/issues/207#issuecomment-78228468)

**Comment by steelbrain** [comment]
Bump. It would be awesome to have it done before 1st of next month, 'cause I am gonna consume remote SSH FS in my packages and I don't wanna fork tree-view.

**Comment by izuzak** [comment]
> Bump. It would be awesome to have it done before 1st of next month,

I don't expect that to happen.

**Comment by steelbrain** [comment]
If I submit a PR providing this functionality, would it be accepted?

**Comment by izuzak** [comment]
> If I submit a PR providing this functionality, would it be accepted?

It's hard to say :8ball: -- this suggestion hasn't gotten any real feedback yet and neither has the PR which you'd create. The answer to your question depends on that feedback.

**Comment by steelbrain** [comment]
Yeah, that's why I bumped it, this feature is critically important for the new Nuclide IDE and Atom-Hack.

**Comment by lee-dohm** [comment]
You submitted an Issue over a holiday weekend, when even us volunteers maybe weren't watching as closely as we normally do because of other commitments. Give people time to get their morning caffeinated beverage, see it and respond :laughing:

The size of this change and not being able to test it without a corresponding provider is going to work against this getting implemented and released quickly.

Also, I feel like there is a lot of equivocation in the solution section of your proposal. Things like:

> The best solution I can think of right now

> The FileSystem Providers would return something like this

> The Tree Provider will provide a protocol:ssh or something similar

It would be hard for me to commit to giving a :+1: on something that doesn't have a committed implementation. This is why PRs are generally easier to discuss for new features. What I'd like to see before I could potentially put my support behind this is:

1. An interface (somewhat similar to what you have above) but including more details about edge cases like:
    1. Would something like this be expected to replace the Archive View?
    1. What happens when a provider doesn't support Unix-like permissions?
    1. What is expected when a provider doesn't support certain operations? For example, if the provider is read-only, would it return `false` for everything? Or some other signal that indicates that operation is not an error because it just isn't supported?
1. Expected UI implementation
    1. Would providers like this become another project root? Or would they potentially also appear inside a file system tree? (On Unix an SSH volume could be mounted inside a Git repository.)
    1. What visual indication would be given to the user that this is an alternate file system?
    1. Would the provider provide some part of that UI, even if it is just an icon to put next to the folder?
    1. How would the user be expected to enter their username and password? Would their credentials be cached? If so, how?
    1. Saving a file is currently considered a synchronous operation. Over a slow network connection this might be problematic. Would saving a file remain synchronous? If writing to a remote file system would become asynchronous, how would you propose the user receive feedback so they know their file is well and truly saved or still in progress?
1. Other impacts to Atom
    1. Would git-utils be supported on these alternate file systems? If so, how?

These questions don't all have to be answered. But they should all be considered in the design of any solution, even if that answer is "we don't know right now, but here are some potential answers".

Programmer's editors are all about managing, reading and writing files. Opening up the question of alternate file systems is a *huge* can of worms. One that I think it would be hugely useful for Atom to tackle at some point in time. Because the effects of it would be so far-reaching, potentially touching virtually everything the application does, it needs to be considered carefully, in my opinion.

**Comment by steelbrain** [comment]
1. An interface

  i. Archive-View, what's that?
  ii. We would have a generic permission shape like `shape('Readable' => bool, 'Writable' => bool)` that is supported everywhere.
  iii. The Promise rejection function will expect an `Error` object, that error will contain a message to be display to the user.

2. Expected UI Implementation

  i. Those Providers will become Project roots and will not be shown inside the file system tree.
  ii. A folder icon that is different from the other folder icons, suggested by the provider.
  iii. The provider will not provide any part of the UI or folder icons. It's contents will behave like a normal tree's contents.
  iv. The credentials won't be cached, The will be prompted to enter them every time they add a folder like that. or If another package opens them programmatically, the package will be responsible for caching them using a configuration file, like I've done in Atom-Hack.
  v. Saving the file can be done synchronously, The user would know about his connection and will be okay with it, but I would be ideal to have it asynchronous and have a progress bar under the Tab.

3. Other Impacts on Atom

  i. Supporting git-utils over a network mount is not ideal. 


and I agree that implementing a virtual file system would include changes in not only tree-view but other components as well as atom core. But once we do so, It will remove a lot of limitations on Atom's package developers as well as users.

**Comment by thedaniel** [comment]
> and I agree that implementing a virtual file system would include changes in not only tree-view but other components as well as atom core

Virtual FS support of some kind is something that would help a lot of users, but for the reason quoted above (i.e. the complexity) as well as the need to reason through the problem thoroughly and test the implementation thoroughly, this is *definitely not* something that we can rush through in a week, and not something that can realistically be in the public Atom 1.0 API.

Performance on network drives etc. is important though, so let's keep working out a solution. Just understand that leading up to the 1.0 release, any change to the public API is off the table and structural changes that will take a lot of effort to test and analyze for performance are also unlikely to be merged.

**Comment by steelbrain** [comment]
It's okay if we can't do this in a week, it was ideal however if it could.

**Comment by pgnd** [comment]
I'm still too new to atom to know if, where & how to implement remote-service-pluggable tree-view in atom.

My $0.02 is that -- of the broad variety of IDEs, etc out there that have tackled the problem, I've only found one that does it thoroughly, flexibly and well (from a user perspective; saying nothing about the code).  That's Eclipse's 'TM/Target Management' (formerly 'RSE/ Remote System Explorer').

this might be an interesting read:  Target Management Home (https://eclipse.org/tm/)

It's already very thorough and well thought out, including a variety of use cases, and (their) answers to some of the questions above.

and, of course, it's resulted in a project/functionality with years of successful operation under its belt.

**Comment by rafaelmaiolla** [comment]
Any chance to use Directory provider to get the entries from it?

**Comment by mgrenier** [comment]
@steelbrain @lee-dohm I'm the original author of [remote-ftp](https://github.com/mgrenier/remote-ftp) package and I've now commited to work on an alternative with Atom's Directory providers.

Last year I talked with an engineer of Facebook Nuclide about their implementation, [nuclide-remote-connection](https://github.com/facebook/nuclide/blob/master/pkg/nuclide-remote-connection/). Turns out they also forked a TreeView-like.

Worked on this over the weekend just to find out that the TreeView is still bound to the file system. Could'nt the tree-view be made compatible with [Directory](https://atom.io/docs/api/v1.8.0/Directory) and [File](https://atom.io/docs/api/v1.8.0/File) instead of the file system? These two we can provide an interface with the `atom.directory-provider` API :
```js
atom.packages.serviceHub.provide('atom.directory-provider', '0.1.0', RemoteDirectoryProvider);
```
Where `RemoteDirectoryProvider` is my provider for FTP/FTPS/SFTP directories.

Subbed to this issue in hope that get more attention in the future.

**Comment by green-coder** [comment]
I would like to bump this issue.

I used to program with NetBeans RCP using Java and its TreeView was really a **generic** tree view where contributors would provide a subclass of `Entry` and `Entry.Children` which had the responsability of determining the type of the nodes, (therefore) the list of actions that could run on it from a contextual menu, and provide their children.

That would be really nice to have something like that in Atom.

My current use case is to show a virtual FS of a project which exists only on a remote server, for a collaboration package. I think I will have to clone TreeView, somehow.

**Comment by mgrenier** [comment]
@green-coder In my spare time I've been hacking my way into this problem. As of right now I've [implemented a prototype](https://github.com/mgrenier/atom-tree-view-vfs) that works with the local file system, AWS S3 bucket and FTP.

A lot needs to be done to complete the package and I haven't found time to work on it lately.
